File: example/learning-agent/java/src/main/java/org/aiddl/example/learning_agent/functions/SampleDataExtractor.java
package org.aiddl.example.learning_agent.functions;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;

import org.aiddl.core.java.interfaces.ConfigurableFunction;
import org.aiddl.core.java.representation.CollectionTerm;
import org.aiddl.core.java.representation.Substitution;
import org.aiddl.core.java.representation.Term;
import org.aiddl.common.java.planning.PlanningTerm;
import org.aiddl.core.java.function.FunctionRegistry;
import org.aiddl.core.java.tools.LockableList;
import org.aiddl.core.java.tools.Logger;

public class SampleDataExtractor implements ConfigurableFunction {

	Random rng;
	
	String name = "ExtractDataFromSamples";
	int verbose = 0;
	
	@Override
	public void configure(Map<Term, Term> settings, FunctionRegistry fReg) {
		this.verbose = settings.getOrDefault(Term.sym("verbose"), Term.integer(0)).getIntValue();
		rng = new Random();
	}

	Term var = Term.anonymousVar();
	private final static Term COLLECTED = Term.sym("collected"); 
	private final static Term DATA = Term.sym("data");
	private final static Term SAMPLE = Term.sym("sample");
	private final static Term EMPTY = Term.sym("empty");
	private final static Term MATCH = Term.keyVal( Term.tuple(COLLECTED, Term.tuple(DATA, Term.var("L"), Term.var("C"))), Term.sym("true"));
	private final static Term MATCH_SAMPLE = Term.keyVal( Term.tuple(SAMPLE, Term.anonymousVar()), Term.anonymousVar());
	private final static Term DATA_KEY = Term.tuple(Term.var("L"), Term.var("C"));
		
	@Override
	public Term apply( Term args ) {
		CollectionTerm state = (CollectionTerm) args.get(Term.sym("state"));
		CollectionTerm model = (CollectionTerm) args.get(Term.sym("model"));
		CollectionTerm data  = (CollectionTerm) args.get(Term.sym("data"));
		
		Set<Term> s_updated = new LinkedHashSet<>();
		
		List<Term> new_data = new ArrayList<>();
		for ( Term p : state ) {
			Term p_sv = p.getKey();
			Term p_a = p.getValue();
			Term sva = Term.keyVal(p_sv, p_a);
			Substitution s = MATCH.match(sva);
			
			if ( s != null ) {
				Term key = DATA_KEY.substitute(s);
				Term observed = model.get(key);
				new_data.add(Term.list(key.get(0), key.get(1), observed));
			} else if ( MATCH_SAMPLE.match(p) != null ) {
				s_updated.add(Term.keyVal(p_sv, EMPTY));
			} else {
				s_updated.add(p);
			}
		}
		
		if ( verbose >= 1 ) {
			Logger.msg(name, "Extracted " + new_data.size() + " samples.");
			if ( verbose >= 2 ) {
				for ( Term t : new_data ) {
					Logger.msg(name, "  " + t.toString() );
				}
			}
		}
			
		data.addAllTo(new_data);
//		new_data.addAll(data.getCollection());
		
		LockableList rList = new LockableList();
		rList.add(Term.keyVal(DataKey, Term.list(new_data)));
		rList.add(Term.keyVal(StateKey, Term.set(s_updated)));
		return Term.list( rList );
	}
	
	private static final Term DataKey = Term.sym("new-data");
	private static final Term StateKey = Term.sym("new-state");
}

File: example/learning-agent-scala/src/main/scala/org/aiddl/example/learning_agent_scala/function/SampleDataExtractor.scala
package org.aiddl.example.learning_agent_scala.function

import org.aiddl.core.scala.function.{Function, Verbose}
import org.aiddl.core.scala.representation.{Bool, CollectionTerm, KeyVal, ListTerm, Num, SetTerm, Sym, Term, Tuple, Var}
import org.aiddl.common.scala.planning.PlanningTerm
import org.aiddl.core.scala.util.ComboIterator

import java.util.logging.Level

class SampleDataExtractor(val hiddenModel: Term) extends Function with Verbose {
  this.logConfig(level=Level.FINE)
  private val DataKey = Sym("new-data")
  private val StateKey = Sym("new-state")

  private val COLLECTED = Sym("collected")
  private val DATA = Sym("data")
  private val SAMPLE = Sym("sample")
  private val EMPTY = Sym("empty")
  private val MATCH = KeyVal(Tuple(COLLECTED, Tuple(DATA, Var("L"), Var("C"))), Bool(true))
  private val RESET_COL = KeyVal(Tuple(COLLECTED, Tuple(DATA, Var("L"), Var("C"))), Bool(false))
  private val MATCH_SAMPLE = KeyVal(Tuple(SAMPLE, Var()), Var())
  private val DATA_KEY = Tuple(Var("L"), Var("C"))

  override def apply(x: Term): Term = {
    val state = x(Sym("state")).asCol
    val model = x(Sym("model")).asCol
    val data = x(Sym("data")).asList

    val (newData, nextState) = this(state, data)
    ListTerm(
      KeyVal(DataKey, data),
      KeyVal(StateKey, state)
    )
  }

  def apply(state: CollectionTerm, data: ListTerm): (ListTerm, SetTerm) = {
    var s_updated: Set[Term] = Set.empty
    var new_data: List[Term] = Nil

    for (p <- state) {
      val p_sv = p.asKvp.key
      val p_a = p.asKvp.value
      val sva = KeyVal(p_sv, p_a)
      //println(p)
      MATCH unify sva match {
        case Some(s) => {
          //println("MATCH: $s")
          val key = DATA_KEY \ s
          val observed = hiddenModel(key)
          new_data = ListTerm(key(0), key(1), observed) :: new_data
          s_updated = s_updated + RESET_COL \ s
        }
        case None => {
          //println("NO MATCH")
          MATCH_SAMPLE unify p match
            case Some(s) => s_updated = s_updated + KeyVal(p_sv, EMPTY)
            case None => s_updated = s_updated + p


        }
      }
    }

    this.logger.info("Extracted " + new_data.size + " samples.")
    for (t <- new_data) {
      this.logger.fine("  " + t.toString)
    }

    (data.addAll(ListTerm(new_data)), SetTerm(s_updated))
  }
}
